File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {

        int maxElement = *max_element(piles.begin(), piles.end());

        int start = 1;
        int end = maxElement;

        int res = -1;

        while (start <= end) {
            int mid = start + (end - start) / 2;

            long long totalTime = 0;
            for (auto& p : piles) {
                totalTime += ceil((double)p / mid);
            }

            if (totalTime <= h) {
                res = mid;
                end = mid - 1;
            }

            else {
                start = mid + 1;
            }


        }

        return res;



        
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:

        maxPile = max(piles)
        total = sum(piles)
        l = 1
        r = maxPile

        while l <= r:
            m = l + (r - l) // 2
            time = 0
            for pile in piles:
                time += math.ceil(pile / m)

            if time <= h:
                r = m - 1
            
            elif time > h:
                l = m + 1
        
        return l


        